ingestion-beam/src/main/java/com/mozilla/telemetry/AmplitudePublisher.java

```diff
@@ -0,0 +1,154 @@
+import org.apache.beam.sdk.transforms.Flatten;
+import org.apache.beam.sdk.transforms.GroupIntoBatches;
+import org.apache.beam.sdk.transforms.WithKeys;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionList;
+import org.joda.time.Duration;
+
+/**
+ * Get event data and publish to Amplitude.
+ */
+public class AmplitudePublisher extends Sink {
+
+  /**
+   * Execute an Apache Beam pipeline.
+   *
+   * @param args command line arguments
+   */
+  public static void main(String[] args) {
+    run(args);
+  }
+
+  /**
+   * Execute an Apache Beam pipeline and return the {@code PipelineResult}.
+   *
+   * @param args command line arguments
+   */
+  public static PipelineResult run(String[] args) {
+    registerOptions(); // Defined in Sink.java
+    final AmplitudePublisherOptions.Parsed options = AmplitudePublisherOptions
+        .parseAmplitudePublisherOptions(PipelineOptionsFactory.fromArgs(args).withValidation()
+            .as(AmplitudePublisherOptions.class));
+    return run(options);
+  }
+
+  /**
+   * Execute an Apache Beam pipeline and return the {@code PipelineResult}.
+   */
+  public static PipelineResult run(AmplitudePublisherOptions.Parsed options) {
+    final Pipeline pipeline = Pipeline.create(options);
+    final List<PCollection<PubsubMessage>> errorCollections = new ArrayList<>();
+
+    PCollection<PubsubMessage> messages = pipeline //
+        .apply(options.getInputType().read(options)) //
+        .apply(ParseUri.of()).failuresTo(errorCollections) //
+        .apply(FilterByDocType.of(options.getAllowedDocTypes(), options.getAllowedNamespaces()));
+
+    // Random sample.
+    if (options.getRandomSampleRatio() != null) {
+      final Double ratio = options.getRandomSampleRatio();
+      messages //
+          .apply("SampleBySampleIdOrRandomNumber", Filter.by(message -> {
+            message = PubsubConstraints.ensureNonNull(message);
+            String sampleId = message.getAttribute("sample_id");
+            return RandomSampler.filterBySampleIdOrRandomNumber(sampleId, ratio);
+          })).apply("RepublishRandomSample", options.getOutputType().write(options));
+    }
+
+    final Map<String, String> apiKeys = readAmplitudeApiKeysFromFile(options.getApiKeys());
+
+    PCollection<KV<String, Iterable<AmplitudeEvent>>> events = messages
+        .apply(DecompressPayload.enabled(options.getDecompressInputPayloads())
+            .withClientCompressionRecorded())
+        .apply("LimitPayloadSize", LimitPayloadSize.toMB(8)).failuresTo(errorCollections) //
+        .apply("ParsePayload", ParsePayload.of(options.getSchemasLocation())) //
+        .failuresTo(errorCollections) //
+        .apply(ParseUserAgent.of()) //
+        .apply(NormalizeAttributes.of()) //
+        .apply(SanitizeAttributes.of(options.getSchemasLocation())) //
+        .apply("AddMetadata", AddMetadata.of()).failuresTo(errorCollections) //
```

Comment:
Might not be needed when reading from republisher

Reply:
might need to be de-compressed?

Reply:
> Might not be needed when reading from republisher

Yes, that's not needed. Republisher reads from decoded topic, see diagram on https://mozilla.github.io/gcp-ingestion/architecture/overview/

Reply:
The input to contextual services is actually all compressed if you look at the counter metric https://console.cloud.google.com/dataflow/jobs/us-west1/2025-02-25_16_51_25-8171389163203304892;graphView=0?project=moz-fx-data-beam-prod-11f7

I think the write to pubsub does the compression so `DecompressPayload` is still needed
https://github.com/mozilla/gcp-ingestion/blob/fa98ac0c8fa09b5671a961062e6cf0985ec48b0e/ingestion-beam/src/main/java/com/mozilla/telemetry/options/SinkOptions.java#L264-L266

Reply:
You're right, I agree we need to decompress.